### ISSUE TITLE
Separate `quotes` rule for Object keys

### ISSUE THREAD

**Issue by palanik** [eslint/eslint#3484]
Currently, the `quotes` rule is applied to all strings.

Although, JS supports both single, double or no quotes for key names in Objects, [JSON spec.](http://json.org/) explicitly states to use double quotes for keys. 

To be consistent with JSON spec., I'd like to have my keys in double quotes, while other strings in single quotes.

I propose a new rule specifically for handling quotes in Object keys - `key-quotes`. 
By default, it could very well inherit the `quotes` rule.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by IanVS** [comment]
Are you saying you want double quotes in your `.json` files, or in all of your object keys in `.js` files?

**Comment by michaelficarra** [comment]
Why follow JSON format in a JS file?

**Comment by palanik** [comment]
@IanVS I want to have double quotes for object keys in `.js` files.

@michaelficarra I understand there is no need to follow JSON format in a JS file. 

Additional quote rule for keys is a nice to have feature, allowing convenient transfer of content between `.json` & '.js` and vice-versa, without change.

**Comment by nzakas** [comment]
We can't create a new rule for this because it would conflict with the `quotes` rule. We could explore adding an option to the `quotes` rule that lets you specify different a setting based on the context, though.

**Comment by palanik** [comment]
Sounds good.

**Comment by nzakas** [comment]
@palanik do you have a suggestion for what that might look like?

**Comment by palanik** [comment]
Perhaps, it makes sense to add the option to `quote-props` instead of `quotes` rule.
```
"quote-props": [2, "always", "double"]
```

**Comment by nzakas** [comment]
We can't do that without removing the check for object properties from `quotes`, which would be a breaking change. Otherwise, the two rules would overlap and you'd get contradictory advice.

**Comment by ilyavolodin** [comment]
Can we add `exceptions` option that can either be `object` or/and `array`? This would be non-breaking change, since the default would still match existing behavior. Or do we just say that since there hasn't been any additional interest in this in over half-a-year, we are not going to do this?

**Comment by nzakas** [comment]
@ilyavolodin can you show an example?

**Comment by ilyavolodin** [comment]
```js
"quote-props": [2, "always", { "exceptions": ["object"] }]
```
Don't take this proposal as me wanting to champion this feature though. I'm not convinced this is a necessary enhancement. Just wanted to move discussion one way or another (it's been sitting around without resolution for a long time).

**Comment by nzakas** [comment]
I'm thinking this is best off as a custom rule for now. I don't see a big need for this in core.
